_posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer.md

```diff
@@ -0,0 +1,90 @@
+
+That looks like this:
+
+![Custom HTTP Basic authorizer](/assets/posts/2018-06-13-Serverless-Basic-Authentication-Custom-Authorizer/api-gateway-http-basic-authorizer.png)
+
+You can find the code for the custom authorizer on [Github](https://github.com/svdgraaf/serverless-basic-authentication/blob/master/basic_auth.py)
+
+```python
+# The custom authorizer gets the Authorization header in the incoming event
+# the contents of the header will be in the authorizationToken key, eg:
+# {
+#     "type":"TOKEN",
+#     "authorizationToken":"Basic Zm1vYmFy1ndNQ1loUGE4TUs5SlFWcGU3dVRqWDVGOEY1MUJXa0Q0YVVGZUI2MnQ=",
+#     "methodArn":"arn:aws:execute-api:<regionId>:<accountId>:<apiId>/<stage>/<method>/<resourcePath>"
+# }
+
+def basicAuth(event, context):
+    authorizationToken = event['authorizationToken']
+
+    # we fetch the base64 encoded token from the header
+    b64_token = authorizationToken.split(' ')[-1]
+
+    # decode the base64 encoded header value
+    username, token = base64.b64decode(b64_token).decode("utf-8").split(':')
+
+    # check if the given api key actually exists
+    client = boto3.client('apigateway')
+    response = client.get_api_keys(nameQuery=username, includeValues=True)
+
+    # if no keys found, deny access
+    if len(response['items']) != 1:
+        print("Couldn't find key")
+        raise Exception('Unauthorized!')
+
+    # if the key value does not match, deny access
+    if response['items'][0]['value'] != token:
+        print("Key value mismatch")
+        raise Exception('Unauthorized!!')
+
+    # All is well, return a policy which allows this user to access to this api
+    # this call is cached for all authenticated calls, so we need to give
+    # access to the whole api. This could be done by having a policyDocument
+    # for each available function, but I don't really care :)
+    arn = "%s/*" % '/'.join(event['methodArn'].split("/")[0:2])
+
+    authResponse = {
+        'principalId': username,
+        'usageIdentifierKey': token,
+        'policyDocument': {
+            'Version': '2012-10-17',
+            'Statement': [{
+                'Action': 'execute-api:Invoke',
+                'Effect': 'Allow',
+                'Resource': arn
+            }]
+        }
+    }
+    print("Authentication response: %s" % authResponse)
+
+    return authResponse
+```
+
+The real magic is in the response code. In the AWS documentation this is fairly poorly documented but the magic is in the `usageIdentifierKey` return value. When this key is present. And Api Gateway is configured to get the key value from the custom authorizer, Api Gateway will use the value of this key as the api key. It needs to contain the *value* of the key, not the name, id or whatever, it's the actual value of the key (took me a while to figure that one out).
```

Comment:
"When this key is present. And Api Gateway"

-> 

"When this key is present and Api Gateway"